Add per-site Xrootd statistics to the HTCondor monitoring.

bbockelm opened the pull request:
Per popular demand, this will allow us to see per-site bytes read and read time, aggregated by site (as identified by the remote server; may be different from the CMS site name).

@osschar

bbockelm committed:
Add per-site Xrootd statistics to the HTCondor monitoring.

Per popular demand, this will allow us to see per-site bytes read and
read time, aggregated by site (as identified by the remote server;
may be different from the CMS site name).

cmsbuild commented:
A new Pull Request was created by  @bbockelm (Brian Bockelman) for master.

It involves the following packages:

FWCore/Framework
FWCore/Services
Utilities/XrdAdaptor


@cmsbuild, @smuzaffar, @Dr15Jones can you please review it and eventually sign? Thanks.
@Martin-Grunewald, @wddgit this is something you requested to watch as well.
@davidlange6 you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pr-code-checks/PR-20143/73

Code check has found code style and quality issues which could be resolved by applying a patch in https://cmssdt.cern.ch/SDT/jenkins-artifacts/pr-code-checks/PR-20143/73/git-diff.patch

In future, you can run `scram build code-checks` to apply code checks

davidlange6 commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 98ecd46993a5b6fe1aa15087b3e6208ade53df3d
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20143/22249/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/56f2f723d8eee5e34bf7ce5ac2e6871fdfa7886d
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20143/22249/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20143/22249/git-merge-result

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20143/22249/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 25
 - DQMHistoTests: Total histograms compared: 2600817
 - DQMHistoTests: Total failures: 250
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2600388
 - DQMHistoTests: Total skipped: 179
 - DQMHistoTests: Total Missing objects: 0
 - Checked 103 log files, 14 edm output root files, 25 DQM output files

bbockelm committed:
Suggested changes from code checks utility.

cmsbuild commented:
Pull request #20143 was updated. @cmsbuild, @smuzaffar, @Dr15Jones can you please check and sign again.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pr-code-checks/PR-20143/89

Dr15Jones commented on Utilities/XrdAdaptor/src/XrdStatistics.h:
```diff
@@ -41,8 +46,16 @@ class XrdStatisticsService
+    // Returns a mapping of "site name" to transfer statistics.  The "site name" is
+    // as self-identified by the Xrootd host; may not necessarily match up with the
+    // "CMS site name".
+    std::map<std::string, CondorIOStats> condorUpdate();
```
A `std::vector<std::pair<std::string, CondorIOStats>` would be more memory efficient and probably faster to fill.

Dr15Jones commented on Utilities/XrdAdaptor/src/XrdStatistics.cc:
```diff
@@ -44,6 +44,26 @@ void XrdStatisticsService::postEndJob()
+    XrdSiteStatisticsInformation *instance = XrdSiteStatisticsInformation::getInstance();
+    if (!instance) {return result;}
+
+    std::lock_guard<std::mutex> lock(instance->m_mutex);
```
For future: exposing a lock is a poor interface. We should rethink this service.

Dr15Jones commented on FWCore/Framework/src/defaultCmsRunServices.cc:
```diff
@@ -23,7 +23,8 @@ namespace edm {
                                               "StatisticsSenderService",
-                                              "CondorStatusService"};
+                                              "CondorStatusService",
+                                              "XrdAdaptor::XrdStatisticsService"};
```
I'm not overly happy about adding more default services to jobs where the framework itself does not need those services to operate.

davidlange6 replied:
hi @dr15jones - perhaps you need two lists here - one for the things the FWK needs and one for the things that CMS wants by default?

Dr15Jones commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: c1f3f5a7a09d42653b074fa3051ed03ec44380ee
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20143/22325/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/7d6f190a9a22b134851a160272dc06c46c1a5063
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20143/22325/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20143/22325/git-merge-result

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20143/22325/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 26
 - DQMHistoTests: Total histograms compared: 2718193
 - DQMHistoTests: Total failures: 285
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2717719
 - DQMHistoTests: Total skipped: 189
 - DQMHistoTests: Total Missing objects: 0
 - Checked 107 log files, 14 edm output root files, 26 DQM output files

Dr15Jones commented:
@bbockelm waiting on a reply to my comments

bbockelm committed:
Use std::vector instead of std::map for providing statistics.

We don't need random access to the map as we only use it as a list
of elements.  Hence, we can use the simpler in-memory structure.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
Pull request #20143 was updated. @cmsbuild, @smuzaffar, @Dr15Jones can you please check and sign again.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/PR-20143/511

bbockelm commented:
@Dr15Jones - I tackled the item that had a concrete action item; however, it's not clear from your other comments what you'd like me to do (or if they're just notes for the future).

davidlange6 commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 7a0484ba2bdb017b032a2ff8c8728d3939ad31a9
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20143/22745/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/b9e49920ac6fb7c8b41a37df4151eae9de6776ea
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20143/22745/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20143/22745/git-merge-result

cmsbuild commented:
Comparison job queued.

Dr15Jones commented:
+1
We need to keep an eye on this since we are now causing all test configurations of cmsRun to force loading of xrootd client.

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @davidlange6, @slava77, @smuzaffar (and backports should be raised in the release meeting by the corresponding L2)

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20143/22745/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 26
 - DQMHistoTests: Total histograms compared: 2656498
 - DQMHistoTests: Total failures: 205
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2656104
 - DQMHistoTests: Total skipped: 189
 - DQMHistoTests: Total Missing objects: 0
 - Checked 107 log files, 14 edm output root files, 26 DQM output files

davidlange6 commented:
+1
